feat(rest.packages.provider): add endpoint for Package Descriptor retrieval

mattdibi opened the pull request:
This PR follows #4894 and adds the endpoint for Eclipse Marketplace's Package Descriptor retrieval.

Can be tested using:

```bash
export ADDRESS=<addr>
export USERNAME=<user>
export PASSWORD=<pass>
```

and then:

```bash
curl -X PUT -k -u $USERNAME:$PASSWORD \
       -H 'Content-Type: application/json' \
       -d '{ "url": "http://marketplace.eclipse.org/marketplace-client-intro?mpc_install=5514714" }' \
       https://$ADDRESS/services/deploy/v2/_packageDescriptor
```

**Roadmap**:
- [X] Contribute `deploy/v2` in its current form (#4868)
- [X] Add endpoint for Deployment Package upload using `multipart/form-data` (#4894)
- [ ] Add Depoloyment Package descriptor download and checking to `install` endpoint (#4934)

### Changelog

I mainly repurposed the code in:

- https://github.com/eclipse/kura/blob/d74484bee2be0501f0d107ff01deec793e6ef384/kura/org.eclipse.kura.web2/src/main/java/org/eclipse/kura/web/client/ui/packages/PackagesPanelUi.java#L592
- https://github.com/eclipse/kura/blob/d74484bee2be0501f0d107ff01deec793e6ef384/kura/org.eclipse.kura.web2/src/main/java/org/eclipse/kura/web/client/ui/packages/PackagesPanelUi.java#L549
- https://github.com/eclipse/kura/blob/9c91aa2848f29aad0ff8fbda55825e1a755053c1/kura/org.eclipse.kura.web2/src/main/java/org/eclipse/kura/web/server/GwtPackageServiceImpl.java#L146
- https://github.com/eclipse/kura/blob/9c91aa2848f29aad0ff8fbda55825e1a755053c1/kura/org.eclipse.kura.web2/src/main/java/org/eclipse/kura/web/server/GwtPackageServiceImpl.java#L239

Creating a new POJO (i.e. `MarketplacePackageDescriptor`) for representing the info contained in the downloaded descriptor and basically using everything in the `rest.packages.provider`.

mattdibi committed:
feat: add MarketplacePackageDescriptor.java

mattdibi committed:
feat: add barebone getMarketplacePackageDescriptor method implementation

WIP

mattdibi committed:
feat: add getMarketplacePackageDescriptor to DeploymentAgentService interface

mattdibi committed:
feat: add isEclipseMarketplaceUrl check

WIP

mattdibi committed:
style: fix variable name

mattdibi committed:
feat: add quick&dirty descriptor check endpoint

STILL WIP: we should at least have a dedicated request type instead of
recycling InstallRequest

mattdibi committed:
feat: return json object

mattdibi committed:
debug: add debugging log

mattdibi committed:
feat: actually perform check on input url

mattdibi committed:
feat: add proper Kura version retrival

mattdibi committed:
refactor: move instantiation

mattdibi committed:
refactor: use proper exceptions

mattdibi committed:
style: add copyright header

mattdibi committed:
feat: use QueryParam for passing the URL to be checked

mattdibi committed:
feat: provide equals and hashCode overrides

mattdibi committed:
fix: do not log user-controlled data

mattdibi committed:
fix: replace character class by the character itself

mattdibi committed:
test: add unit tests for MarketplacePackageDescriptor

mattdibi committed:
style: add copyright header

mattdibi committed:
style: refactor into gerkhin style

mattdibi committed:
refactor: move given/when/then methods in their sections

mattdibi committed:
refactor: use the builder for gods sake

mattdibi committed:
style: add forgotten about files

mattdibi committed:
test: add DeploymentRestServiceUnitTest for marketplace descriptor

mattdibi committed:
refactor: move URL checks outside DeploymentAgentService to allow for testing

mattdibi committed:
test: update unit tests

mattdibi committed:
refactor: move method into its section

mattdibi committed:
test: add integration tests

mattdibi committed:
fix: remove commented out code

mattdibi committed:
test: add some more tests

mattdibi commented:
Setting this as ready for review since the only thing missing are tests at this point

mattdibi committed:
test: add test case for failing getMarketplacePackageDescriptor call

mattdibi committed:
docs: add Javadocs for DeploymentRestService

mattdibi committed:
docs: add javadocs for DeploymentAgentService

mattdibi commented:
@nicolatimeus I'd like to have some feedback on the following topics:
- **Query parameters**: is it ok to pass the `url` as a query parameter? In other APIs we used to `POST` a JSON object, but there's a constraint due to how the code is also used for the MQTT side of things (we don't want to encode data in the URL due to the additional computational burden in the cloud side of things). Here we don't have the same constraint and I think the `GET` method is more correct in this specific instance. If this is not ok and we want to keep the API consistent, I can change it in little to no time so no worries. **Update** -> We decided to go with `PUT` due to consistency and security concerns.
- **DeploymentAgentImpl method implementation**: I split [the method](https://github.com/eclipse/kura/blob/39a6a93a1c7b1312fbb10ba956e29d63ba4052ef/kura/org.eclipse.kura.deployment.agent/src/main/java/org/eclipse/kura/deployment/agent/impl/DeploymentAgent.java#L266) so that it just accepts the url without checking it belongs to the Eclipse Marketplace, this allows for proper unit testing (if the check is performed inside the `getMarketplacePackageDescriptor` I wouldn't be able to point the method to a mock server for example... which I'm doing in my draft unit tests). The checking is delegated to the caller. Is this ok?
- **DeploymentAgent tests**: I had some trouble in making the added dependencies work. I had to [add them both as `dependencies`](https://github.com/eclipse/kura/blob/0563e7c39d8fc405b21ac02feb2c89287415ad6a/kura/test/org.eclipse.kura.deployment.agent.test/pom.xml#L34) in the pom and [download them inside a `lib` folder](https://github.com/eclipse/kura/blob/0563e7c39d8fc405b21ac02feb2c89287415ad6a/kura/test/org.eclipse.kura.deployment.agent.test/pom.xml#L59). Some help there is appreciated.

**Note**: [Mockserver](https://www.mock-server.com/) that I used for the tests uses [Apache License 2.0](https://github.com/mock-server/mockserver/blob/master/LICENSE.md) therefore it should be compatible.

mattdibi committed:
test: add DeploymentAgentTest:getMarketplacePackageDescriptor tests (#11)

mattdibi committed:
ci(debug): set envvar to run testcontainers with podman

mattdibi committed:
refactor: user raw Mockserver instead of going through Testcontainers

mattdibi committed:
refactor: remove unnecessary dependencies

mattdibi committed:
ci: roll-back changes

mattdibi committed:
test: use random free port

mattdibi committed:
test: add null URL test

mattdibi committed:
test: reset MockServer status after each test

mattdibi committed:
feat: GET -> PUT

mattdibi committed:
test: GET -> PUT

mattdibi committed:
refactor(web2): modify UI so that it uses the newly introduced method

mattdibi committed:
build: downgrade mockserver 5.15.0 -> 5.14.0

mattdibi commented:
Dependency introduced are ✅ 

<img width="1335" alt="image" src="https://github.com/eclipse/kura/assets/22748355/7cb11b86-497d-40c4-9fb0-7ce54653f6e6">

mattdibi committed:
build: I'm a dum dum

nicolatimeus commented on kura/org.eclipse.kura.deployment.agent/src/main/java/org/eclipse/kura/deployment/agent/DeploymentAgentService.java:
```diff
@@ -99,6 +99,18 @@ public interface DeploymentAgentService {
+    /**
+     * Provides the Eclipse Marketplace Package Descriptor information of the deployment package identified by URL
+     * passed as query parameter.
+     *
```
Since this is a new method in an exported package and it has a javadoc, please also add `@since 1.4.0` annotation. (1.4.0 is the bundle version that introduces the new method)

mattdibi replied:
Addressed in 280d140e1ec4442e1c89dfeea17c83967b651184

nicolatimeus commented:
Please update the exported package version to 1.1.0 here https://github.com/mattdibi/kura/blob/73adee3550e2d4febf49bfe16abe7207aa4078dc/kura/org.eclipse.kura.deployment.agent/META-INF/MANIFEST.MF#L9

nicolatimeus commented on kura/org.eclipse.kura.web2/src/main/java/org/eclipse/kura/web/server/GwtPackageServiceImpl.java:
```diff
@@ -126,136 +114,34 @@ public void uninstallDeploymentPackage(GwtXSRFToken xsrfToken, String packageNam
-
-            mpUrl = new URL(String.format(MARKETPLACE_URL, nodeId));
-            connection = (HttpsURLConnection) mpUrl.openConnection();
-            connection.setSSLSocketFactory(this.sslManagerService.getSSLSocketFactory());
```
Currently the console has a configuration property that allows to specify the `SslManagerService` that is used to fetch deployment packages from the marketlplace, this change would make it use the SslManagerService configured in the deployment agent instead.. We could add another `getMarketplacePackageDescriptor(String, SslManagerService)` to the DeploymentAgent that allows to specify the `SslManagerService` to restore the old behaviour

mattdibi replied:
Addressed in 08eca41ab8e2c7d638e0938176e0058bf65414e1

nicolatimeus commented:
Please also update copyright headers in all changed files

nicolatimeus commented:
> Please update the exported package version to 1.1.0 here https://github.com/mattdibi/kura/blob/73adee3550e2d4febf49bfe16abe7207aa4078dc/kura/org.eclipse.kura.deployment.agent/META-INF/MANIFEST.MF#L9

We should also update the import declarations of this package to use version range `[1.1,2.0)` in the bundles that use the new methods (console)

mattdibi committed:
docs: add "since" annotation for newly introduced method

mattdibi committed:
style: update copyright headers

mattdibi committed:
chore: deployment agent service version bump to 1.1.0

mattdibi commented:
> Please update the exported package version to 1.1.0 here https://github.com/mattdibi/kura/blob/73adee3550e2d4febf49bfe16abe7207aa4078dc/kura/org.eclipse.kura.deployment.agent/META-INF/MANIFEST.MF#L9

Addressed in d0dee4cd66057235a6a2dc7857d37f9d59db141f

> We should also update the import declarations of this package to use version range [1.1,2.0) in the bundles that use the new methods (console)

Addressed in d0dee4cd66057235a6a2dc7857d37f9d59db141f

> Please also update copyright headers in all changed files

Addressed in 98cbbaa1ff44c66f4c54ac6f13466c79e71b0fd4

mattdibi committed:
feat: add method for overriding SslManagerService used to establish connection

mattdibi committed:
style: restore formatting

mattdibi committed:
style: update copyright headers

mattdibi committed:
style: update copyright header

mattdibi committed:
build: switched to mockserver v 5.15.0

mattdibi commented:
Update to mockserver v 5.15.0. Newly added dependency is ✅ 

```
echo "maven/mavencentral/org.mock-server/mockserver-netty-no-dependencies/5.15.0" | java -jar ~/Downloads/org.eclipse.dash.licenses-1.0.3-20231111.065112-172.jar -
[main] INFO Querying Eclipse Foundation for license data for 1 items.
[main] INFO Found 1 items.
[main] INFO Vetted license information was found for all content. No further investigation is required.
```

```
echo "maven/mavencentral/org.mock-server/mockserver-client-java-no-dependencies/5.15.0" | java -jar ~/Downloads/org.eclipse.dash.licenses-1.0.3-20231111.065112-172.jar -
[main] INFO Querying Eclipse Foundation for license data for 1 items.
[main] INFO Found 1 items.
[main] INFO Vetted license information was found for all content. No further investigation is required.
```

```
maven/mavencentral/org.mock-server/mockserver-client-java-no-dependencies/5.15.0, Apache-2.0 AND LGPL-3.0-only, approved, #11514
maven/mavencentral/org.mock-server/mockserver-netty-no-dependencies/5.15.0, Apache-2.0 AND LGPL-3.0-only AND OFL-1.1 AND (BSD-3-Clause AND MIT), approved, #11515
```

See:
- https://gitlab.eclipse.org/eclipsefdn/emo-team/iplab/-/issues/11514
- https://gitlab.eclipse.org/eclipsefdn/emo-team/iplab/-/issues/11515

mattdibi committed:
docs: remove newly added test deps from NOTICE file
